### PR TITLE
27 support 3d layers in dfm bmi interface

### PR DESCRIPTION
**Pull request by erwin-mulder**

### CODE REVIEW

**Review by vrielin — CHANGES_REQUESTED** [review]
I did not review the changes in detail (yet).. I tried to run these 2 examples

```
p:\dflowfm\maintenance\JIRA\07000-07999\07576\02_test_models\1_lock_cycle-average\
p:\dflowfm\maintenance\JIRA\07000-07999\07576\02_test_models\1_lock_cycle-average_3D\
```
with this branch (and the current head of D-FlowFM main) and both runs end on:
`assert(layers->number_of_layers > 0); in distribute_discharge_over_layers 
`
I think number_of_layers is initialized to 0 but never updated. I can''t find the location where sealock_get_active_layers is called.

In the 3D example there is water_volume specified as exchange item in the dimr_config.xml.
For the 2D example I think a num_layers = 1 and volumes = 1.0 init is needed? Or prevent calling the distribute methods?

**Comment by erwin-mulder** [comment]
Upon reviewing my own code just now, I spotted at least one 'one-off' bug which could result in zero layers when there is in fact one. I'll address this first, and maybe we should assume having at least one layer if nothing is configured in the .ini file(s)? I will have to look into that bit.

File: src/dimr_bmi.c [hunk 1]
```diff
@@ -235,7 +235,9 @@ int update(double dt) {
   printf("ZSF: %s( %g ) called.\n", __func__, dt);
 #endif
   if (dt <= 0) {
-    return DIMR_BMI_FAILURE;
+    // DIMR sends dt == 0 as a first timestep, to singal 'init'.
+    // We will silently ignore this as we are already set up.
+    return DIMR_BMI_OK;
```
Comment (vrielin) [inline src/dimr_bmi.c]: typo: singal = signal. 

Wouldn't it be better to still return a failure for dt < 0?

Comment (erwin-mulder) [reply src/dimr_bmi.c]: Fixed it and sending failure for < 0, ==0 is a bit dangerous as dt is a double, casing to (int) in that case. Hope that works reliably.

**Review by vrielin — APPROVED** [review]
1. I am fine with the code changes.
2. existing 2d test 1_lock_cycle-average\ is still OK.
3. the 3d test 1_lock_cycle-average\_3D finishes correctly, more checking on exact values should be done during testing.

**Pull request merged**